src/Balances.sol

```diff
@@ -0,0 +1,70 @@
+// SPDX-License-Identifier: (MIT or Apache-2.0)
+
+import "../lib/openzeppelin-contracts/contracts/utils/Nonces.sol";
+import "../lib/openzeppelin-contracts/contracts/utils/cryptography/ECDSA.sol";
+pragma solidity ^0.8.19;
+
+contract Balances is Nonces {
+    event Withdrawal(address indexed account, address target, uint256 value);
+
+    mapping(address => uint) public balances;
+    uint public balanceHeld = 0;
+
+    function balanceOf(address account) public view returns (uint) {
+        return balances[account];
+    }
+
+    function availableBalance() public view returns (uint) {
+      return address(this).balance - balanceHeld;
+    }
+
+    function freeUpBalance(uint amount) internal {
+      balanceHeld -= amount;
+    }
+
+    function reserveBalance(uint amount) internal {
+      balanceHeld += amount;
```

Comment:
where `reserveBalance` is called, we're already ensuring we're not reserving more than available. I'll add a comment about this, instead of adding the check (which increases gas)